accepted/2024/repo-level-lock-file.md

```diff
@@ -0,0 +1,84 @@
+﻿# ***Repository Level Lock File***
+
+- Author: [@CEbbinghaus](https://github.com/CEbbinghaus) 
+- Issue: [Nuget/Home#12409](https://github.com/NuGet/Home/issues/12409) 
+
+## Summary
+
+Applications consisting of multiple solutions (using assembly references between projects) cannot control their transitive dependencies by using Nuget. The current lock file is on a per-project basis rather than an application/repository basis. Given a `Directory.packages.props` file, each project could resolve different versions of transitive dependencies, which a single central lock file would solve by allowing consistency between package versions across multiple solutions.
+
+
+## Motivation 
+
+Monorepositories with multiple solutions utilizing assembly references that deploy all their code together as a single application are unable to use Nuget (with or without CPM) due to transitive version conflicts. This is due to NuGet being able to resolve different versions of transitive dependencies for each individual build, as they occur in isolation from one another. A single central lock file allows all transitive dependencies to have locked versions defined at the root, thereby eliminating third-party dependency management software such as Paket.
+
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented, and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real-life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+The `Directory.packages.props` now uses the`<NuGetLockFilePath>[path]</NuGetLockFilePath>` property within a `<PropertyGroup>` to specify its lock file path. It becomes the source of truth for all builds. Project level lock files are no longer required unless a package version is overridden.
```

Comment:
Is there a reason to make the lock file path configurable in the project? AFAIK, most package managers leave it unconfigured, and that way, the package manager wouldn't have to first read the configuration file first in order to install packages, only the lock file(which makes it easier to separate into steps).

Reply:
The package manager would need to read the Directory.packages.props file anyways as it is what enables CentralPackageManagement in the first place? ([relevant docs](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#enabling-central-package-management))

We could get away with setting a property like `<RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>` to enable CPM Lock files and implicitly assume the `<NuGetLockFilePath>` to be `packages.lock.json` but we can also achieve the same with a single property and a check `$(NuGetLockFilePath) = ''`. At the end it will be up to the maintainers to give feedback which way they prefer but the lock file path will need to be configurable in any case.

Reply:
> The package manager would need to read the Directory.packages.props
Yeah, but you can separate that into steps right? I'm imagining a use case of being able to separate it into layers:
You read the configuration file and can generate the lock file in one step, and then you can read the lock file and get the packages in a different step. The reason this could be useful is that you can write a completely separate program to read the lock file if you wanted to(one that assumes that you're using CPM), and you wouldn't have to read the configuration file to figure out where the lock file is. 

I personally can't think of a great reason to not have the lock file at the top level of a repo(or at some fixed path), and I can think of more reasons to avoid being able to configure it(Easier to make external tools to consume the lock files and process them separately, avoiding conflicts between, say, a git submodule with it's own lock file that's configured to generate this lock file, etc.).

Given that, as far as I'm aware, most package managers(mostly python and node ones) that I've used have a fixed path for their global lock files, I thought I'd at least bring it up.

Reply:
You would need a way to define the "top level" location. This allows for that.

Keep in mind that not every repo-project relationship is 1:1, particularly in a monorepo situation where you might want a small handful of "top levels" rather than only one.

Reply:
It should just be in the same folder as the Directory.Packages.props file, the same way that the "top level" of a nodejs project is where the package.json is.

Reply:
Sorry i feel you might have misunderstood. This is exactly the case @purepani the `<NuGetLockFilePath>` specifies the lock file for the CPM rather than each individual project. This will be located relative to the `Directory.Packages.props` at the root of the project.

Now that you mention it we will likely need a property like `RestorePackagesWithLockFile` but specifically to enable CPM lock files but the lockfile will be located at the root NOT under individual projects

Reply:
I see so this property is meant for each individual project rather than just the top level of the projects; that makes way more sense. 
The way pnpm solves this is by symlinking the dependencies into each workspace after doing a `pnpm install` at the top level. If you only wanted some dependencies, you could just install it for that group. In this case, it is just the dependencies and not the lock file, but you could certainly symlink the lock file if you wanted(though the only advantage of that approach would be that you wouldn't have to move to the top level every time you wanted to install a package).